staging/src/k8s.io/client-go/tools/remotecommand/websocket.go

```diff
@@ -0,0 +1,344 @@
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	if conn == nil {
+		panic(fmt.Errorf("websocket connection is nil"))
+	}
+	var streamer streamProtocolHandler
+
+	streamingProto := conn.Subprotocol()
+	klog.V(4).Infof("The subprotocol is %s", streamingProto)
+	if streamingProto != remotecommand.StreamProtocolV5Name {
+		return fmt.Errorf("unsupported streaming protocol: %q", streamingProto)
+	}
+	streamer = newStreamProtocolV4(options)
+
+	creator := newWSStreamCreator(conn)
+	go creator.run()
+	return streamer.stream(creator)
+}
+
+type wsStreamCreator struct {
+	conn      *gwebsocket.Conn
+	connMu    sync.Mutex
+	streams   map[byte]*stream
+	streamsMu sync.Mutex
+}
+
+func newWSStreamCreator(conn *gwebsocket.Conn) *wsStreamCreator {
+	ws := wsStreamCreator{
+		conn:    conn,
+		streams: map[byte]*stream{},
+	}
+
+	go ws.sendPings(PingPeriod) // start heartbeat
+
+	return &ws
+}
+
+func (c *wsStreamCreator) getStream(id byte) *stream {
+	c.streamsMu.Lock()
+	defer c.streamsMu.Unlock()
+	return c.streams[id]
+}
+
+func (c *wsStreamCreator) setStream(id byte, s *stream) {
+	c.streamsMu.Lock()
+	defer c.streamsMu.Unlock()
+	c.streams[id] = s
+}
+
+// CreateStream uses id from passed headers to create a stream over "c.conn" connection.
+// Returns a Stream structure or nil and an error if one occurred.
+func (c *wsStreamCreator) CreateStream(headers http.Header) (httpstream.Stream, error) {
+	streamType := headers.Get(v1.StreamType)
+	id, ok := streamType2streamID[streamType]
+	if !ok {
+		return nil, fmt.Errorf("unknown stream type: %s", streamType)
+	}
+	if s := c.getStream(id); s != nil {
+		return nil, fmt.Errorf("duplicate stream for type %s", streamType)
+	}
+	reader, writer := io.Pipe()
+	s := &stream{
+		headers:   headers,
+		readPipe:  reader,
+		writePipe: writer,
+		conn:      c.conn,
+		connMu:    &c.connMu,
+		id:        id,
+	}
+	c.setStream(id, s)
+	return s, nil
+}
+
+// sendPings starts heartbeat sending a "ping" from the endpoint every "period".
+func (c *wsStreamCreator) sendPings(period time.Duration) {
+	c.connMu.Lock()
+	closedCh := make(chan struct{})
+	closeHandler := c.conn.CloseHandler()
+	c.conn.SetCloseHandler(func(code int, text string) error {
+		close(closedCh)
+		return closeHandler(code, text)
+	})
+	c.connMu.Unlock()
+	t := time.NewTicker(period)
+	defer t.Stop()
+	for {
+		select {
+		case <-closedCh:
+			return
+		case <-t.C:
+			c.connMu.Lock()
+			if err := c.conn.WriteControl(gwebsocket.PingMessage, nil, time.Now().Add(period)); err != nil {
+				klog.V(7).Infof("Websocket Ping failed: %v", err)
+				// Continue, in case this is a transient failure.
+				// c.conn.CloseChan above will tell us when the connection is
+				// actually closed.
+			} else {
+				klog.V(7).Infof("Websocket Ping succeeeded")
+			}
+			c.connMu.Unlock()
+		}
+	}
+}
+
+func (c *wsStreamCreator) run() {
+	readBuffer := make([]byte, remotecommand.WebSocketBufferSize)
```

Comment:
what does this buffer size have to correspond to, and what happens if it differs? everything in this package gets shipped as client libraries which we have a hard time ensuring stay updated, so we have to be really confident in the correctness of the stuff in this package (and/or in our ability to mitigate issues server-side)

Reply:
(the same comment about this being a client package we can't easily ensure is universally updated holds for the other questions aojea had... there seem to be several unresolved threads)

Reply:
Added comment above this `readBuffer` creation describing the subtle problem with incomplete websocket connection reads when the buffer size is not the same as the websocket connection read buffer size (specified in websocket creation).

Reply:
It's pretty confusing to have a comment saying the buffer sizes have to match, and then have the buffer size be explicitly one greater than the constant in `RoundTripper#RoundTrip`:

```go
ReadBufferSize: remotecommand.WebSocketBufferSize + 1,
```

Can this be clarified or reconciled?

Reply:
also, I changed the readbuffersize to `WebSocketBufferSize + 1`, `WebSocketBufferSize`, and `WebSocketBufferSize - 1`, and tests passed with all three, so either the buffer size doesn't matter or we're not exercising the issue caused by the buffer size mismatch in tests.

If it's really important we match, we need a test demonstrating the issue so we know it doesn't creep back in in the future